deltacat/utils/pyarrow.py

```diff
@@ -133,6 +133,24 @@ def slice_table(
     return tables
 
 
+def slice_table_generator(
+        table: pa.Table,
+        max_len: int) -> Generator[pa.Table, None, None]:
+    """
+    Generator for creating 0-copy table slices.
+    """
+    offset = 0
+    records_remaining = len(table)
+    while records_remaining > 0:
```

Comment:
When the max records per file is say 100 (N) but your table has 100000 (M) records, this would yield 1000 smaller pyarrow tables. The 1000 compute iterations could be saved if you just split it into two parts one of size `floor(M/N)*N` and second of `M%N`.

Reply:
The idea here is to save the 1000 compute iterations everytime we add a new entry to remainder table.

Reply:
The slicing only occurs on `stage_delta` if the table being passed in has greater record counts than `max_records_per_output_file`. So by effectively passing in a concatenated table with a guaranteed size of `max_records_per_output_file` (or less than, for the final remainder), we can guarantee that each concatenated table would result in one file written to S3 (and not have any further splitting).

Basically, pre-splitting to `[100, 100, 100, ..., 100, 1]` or just bundling upfront to `[100_000, 1]` doesn't really change the fact that we compute 1000 iterations. It might be a bit nuanced behavior, but pre-splitting is convenient here since we are also slicing and merging table records based on the offset of previous remainders.

Reply:
Correct me if I am wrong, don't we end up calling data catalog multiple times to create new "in-progress" delta for each file instead of calling data catalog once for all the files uploaded during a single `stage_delta` call.

Reply:
Currently, the data catalog API internally slices a given table into groups of `max_records_per_output_file` tables, and performs a single S3 call per table.

https://github.com/ray-project/deltacat/blob/d4b307cae12c8d8a6f6ccbc1fe6169e8a3fd20e5/deltacat/aws/s3u.py#L281

Reply:
I've since revised my changes to do post-splitting instead, since the data catalog API call itself could be expensive and it would be good to reduce the number of these calls as much as we can (we'll still have to live with more than a single stage_delta call per materialize task, if we eagerly materialize).